### PR TITLE
2/n Move Precision Plugin into strategy - move optimizer related logics

### LINKED ISSUE #7324: Precision Plugins should be part of Training Type Plugins
## 🚀 Feature

Precision Plugin should be a property of the training type plugin, not the accelerator. these can overlap, and interleaving sequences of operations across the two in the accelerator can be challenging to get fully right.

cc @justusschock @awaelchli @kaushikb11 @tchaton 

todo:  fill out in more detail

### Motivation

This would make management of things like this easier:
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7271
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7217

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?
1/n Move Precision Plugin into Strategy and update references
**2/n Move optimizer, lr to Strategy**
3/n Simplify and reduce hooks and redundancies

Fixes #7324 which is part of #10416

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/gpu.py [hunk 1]
```diff
@@ -37,12 +37,14 @@ def setup_environment(self) -> None:
...
-        torch.cuda.set_device(self.root_device)
+        if "cuda" not in str(self.training_type_plugin.root_device):
+            raise MisconfigurationException(
+                f"Device should be GPU, got {self.training_type_plugin.root_device} instead"
+            )
+        torch.cuda.set_device(self.training_type_plugin.root_device)
 
     def setup(self, trainer: "pl.Trainer") -> None:
-        self.set_nvidia_flags(trainer.local_rank)
+        self.set_nvidia_flags(getattr(self.training_type_plugin, "local_rank", 0))
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/gpu.py:47]: this can keep using trainer.local_rank as long as we have the trainer reference.

File: pytorch_lightning/lite/wrappers.py [hunk 2]
```diff
@@ -50,18 +50,20 @@ def __init__(self, optimizer: Optimizer, accelerator: Accelerator) -> None:
         self.__class__ = type("Lite" + optimizer.__class__.__name__, (self.__class__, optimizer.__class__), {})
         self._optimizer = optimizer
         self._accelerator = accelerator
+        # TODO refactor to take Strategy as param, API breaking change for Lite? @
+        self._strategy = self._accelerator.training_type_plugin
```
Comment (awaelchli) [inline pytorch_lightning/lite/wrappers.py:54]:
```suggestion
        # TODO (@awaelchli) refactor to take Strategy as param
        self._strategy = self._accelerator.training_type_plugin
```

nope, no problem. 
I anticipated this and we marked _LiteOptimizer as protected. This is not user facing so all good.
I'll handle this in a separate PR!

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 3]
```diff
@@ -47,13 +48,17 @@ def main_params(self, optimizer: Optimizer) -> _PARAMETERS:
...
-            _, accelerator.optimizers = amp.initialize(
-                trainer.lightning_module, accelerator.optimizers, opt_level=self.amp_level
+            strategy = trainer.training_type_plugin
+            _, strategy.optimizers = amp.initialize(
+                trainer.lightning_module, strategy.optimizers, opt_level=self.amp_level
             )
             self._connected = True
         return super().dispatch(trainer)
 
+    @property
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:58]: Let's move properties to the top, after `__init__` and before the methods? Same would apply for other plugins

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 4]
```diff
@@ -47,13 +48,17 @@ def main_params(self, optimizer: Optimizer) -> _PARAMETERS:
...
+    def amp_backend(self) -> Optional[LightningEnum]:
+        return AMPType.APEX
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:60]: note, there is already a backend class attribute, so we should probably use that for now?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 5]
```diff
@@ -66,8 +71,105 @@ def setup_environment(self) -> None:
...
+        self.lr_schedulers = lr_schedulers
+        self.optimizer_frequencies = optimizer_frequencies
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        """Setup plugins for the trainer fit and creates optimizers.
+
+        Args:
+            trainer: the trainer instance
+        """
+        # call super()
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:95]: not needed I believe

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 6]
```diff
@@ -517,7 +517,7 @@ def _initialize_deepspeed_train(self, model):
         self.model = model
 
     @contextlib.contextmanager
-    def model_sharded_context(self) -> Generator[None, None, None]:
+    def model_sharded_context(self) -> Generator:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:520]: I think the annotation is needed, as without it the return type of the generator would be unknown to mypy

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 7]
```diff
@@ -109,14 +109,23 @@ def __init__(
...
         # patch the dataloader creation function with the custom `poptorch.DataLoader`.
         # this violates the intended control flow for the plugins, but since this is experimental, we have chosen
         # to use the simpler solution before adding abstractions to override the `DataLoader` class
         self.lightning_module.trainer._update_dataloader = self._convert_to_poptorch_loader
+        super().setup(trainer)
+
+    def setup_optimizers(self, trainer: "pl.Trainer") -> None:
+        # refactor after move accelerator into strategy @four4fish
+        # RFC: I think set optimizer related logic should be in strategy instead of accelerator.
+        if len(self.optimizers) > 1:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ipu.py:125]:
```suggestion
        # TODO (@four4fish): refactor after move accelerator into strategy
        if len(self.optimizers) > 1:
```
Let's do the discussion on the main #10416 about

> RFC: I think set optimizer related logic should be in strategy instead of accelerator.

File: tests/lite/test_wrappers.py [hunk 8]
```diff
@@ -145,8 +146,10 @@ def test_lite_optimizer_wraps():
 def test_lite_optimizer_steps():
     """Test that the LiteOptimizer forwards the step() and zero_grad() calls to the wrapped optimizer."""
     optimizer = Mock()
-    accelerator = Mock()
+    strategy = Mock()
+    accelerator = Accelerator(strategy)
```
Comment (awaelchli) [inline tests/lite/test_wrappers.py:150]: this is still outdated from a previous verson of your code where the first argument was the strategy, but now it is the precision plugin

```suggestion
    accelerator = Accelerator(training_type_plugin=strategy)
```

File: pytorch_lightning/accelerators/accelerator.py [hunk 9]
```diff
@@ -133,36 +111,13 @@ def lightning_module(self) -> "pl.LightningModule":
...
-        return self.training_type_plugin.root_device
-
     def teardown(self) -> None:
         """This method is called to teardown the training process.
 
         It is the right place to release memory and free other resources.
         """
         self.training_type_plugin.teardown()
 
-    def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:148]: I think it make sense to keep it in Accelerator, will move back

File: pytorch_lightning/accelerators/accelerator.py [hunk 10]
```diff
@@ -80,28 +68,18 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
-        self.setup_training_type_plugin()
-        if not self.training_type_plugin.setup_optimizers_in_pre_dispatch:
-            self.setup_optimizers(trainer)
-        self.setup_precision_plugin()
+        self.training_type_plugin.setup(trainer)
 
     def pre_dispatch(self, trainer: "pl.Trainer") -> None:
         """Hook to do something before the training/evaluation/prediction starts."""
-        self._move_optimizer_state()
+        self.training_type_plugin._move_optimizer_state()
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:75]: I think all these hooks will be moved in 3/n with X_steps

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/plugins/precision/apex_amp.py:58]: Make sense, will do for all files. Will also reorg the function orders in all plugins

File: pytorch_lightning/accelerators/accelerator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:75]: Strategy has overrides for pre_dispatch(), if we move one line, all the subclass need super(). How about let's move it in 3/x together?

File: pytorch_lightning/accelerators/accelerator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:148]: @four4fish couldn't we handle this in a separate PR? this PR is supposed to be for optimizer related logic, but the batch is separate from that.

[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:148]: Make sense! will do

File: pytorch_lightning/accelerators/accelerator.py [hunk 11]
```diff
@@ -195,121 +150,6 @@ def predict_step(self, step_kwargs: Dict[str, Union[Any, int]]) -> STEP_OUTPUT:
...
-        model_ref.optimizer_zero_grad(current_epoch, batch_idx, optimizer, opt_idx)
-
-    def setup_optimizers(self, trainer: "pl.Trainer") -> None:
-        """Creates optimizers and schedulers.
-
-        Args:
-            trainer: the Trainer, these optimizers should be connected to
-        """
-        if trainer.state.fn not in (TrainerFn.FITTING, TrainerFn.TUNING):
-            return
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:246]: for future: we should check this before calling `setup_optimizers` to be more explicit

File: pytorch_lightning/accelerators/cpu.py [hunk 12]
```diff
@@ -29,8 +29,10 @@ def setup(self, trainer: "pl.Trainer") -> None:
             MisconfigurationException:
                 If the selected device is not CPU.
         """
-        if "cpu" not in str(self.root_device):
-            raise MisconfigurationException(f"Device should be CPU, got {self.root_device} instead.")
+        if "cpu" not in str(self.training_type_plugin.root_device):
+            raise MisconfigurationException(
+                f"Device should be CPU, got {self.training_type_plugin.root_device} instead."
+            )
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/cpu.py:35]: do you think these checks should sit in the strategy after the refactor? the accelerator won't have a reference to training type soon.

Comment (four4fish) [reply pytorch_lightning/accelerators/cpu.py:35]: Yeah, it will happen in step3 "move accelerator/device" to strategy with other misconfig checks.
I think it make more sense to do in step3, as when we move all accelerator.setup() login into strategy. There are other misconfi checks like what strategy/precision tpu should use.

File: pytorch_lightning/accelerators/accelerator.py (hunk 11 above)
Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:246]: I thought about it, but ipu has override for this.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 5 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:95]: Was a note to myself to remember call super if there is override in subclass, will remove

File: pytorch_lightning/accelerators/accelerator.py [hunk 13]
```diff
@@ -133,11 +112,6 @@ def lightning_module(self) -> "pl.LightningModule":
         """
         return self.training_type_plugin.lightning_module
 
-    @property
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:136]: Why do we need to move root device in this PR?

File: pytorch_lightning/trainer/trainer.py [hunk 14]
```diff
@@ -1397,7 +1397,7 @@ def _call_setup_hook(self) -> None:
         self.training_type_plugin.barrier("post_setup")
 
     def _call_configure_sharded_model(self) -> None:
-        with self.accelerator.model_sharded_context():
+        with self.training_type_plugin.model_sharded_context():
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1400]: nit: this could be a separate PR

File: pytorch_lightning/accelerators/accelerator.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:136]: From the PR title: `move optimizer related logics`. I believe this could be within another PR.

File: pytorch_lightning/accelerators/accelerator.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:246]: From IPUs, it seems the logic is quite encapsulated and ipu checks the number of optimizers as it currently supports only 1.

```py
    def setup_optimizers(self, trainer: "pl.Trainer") -> None:
        """
        Raises:
            MisconfigurationException:
                If multiple optimizers are provided.
        """
        super().setup_optimizers(trainer)

        if len(self.optimizers) > 1:
            raise MisconfigurationException("IPUs currently only support one optimizer.")
```

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 15]
```diff
@@ -55,7 +57,15 @@ def __init__(
...
         self.device = device
-        self.scaler = scaler
+        self._scaler = scaler
+
+    @property
+    def scaler(self) -> Optional["GradScaler"]:
+        return self._scaler
+
+    @property
+    def amp_backend(self) -> Optional[LightningEnum]:
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/native_amp.py:67]: Neat !

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 16]
```diff
@@ -48,6 +50,14 @@ def connect(
         """Connects this plugin to the accelerator and the training process."""
         return model, optimizers, lr_schedulers
 
+    @property
+    def scaler(self) -> Optional["GradScaler"]:
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/precision_plugin.py:54]: Should we raise a NotImplementedError instead? Same above ?

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 17]
```diff
@@ -157,7 +157,7 @@ def configure_ddp(self) -> None:
             self.model_to_device()
 
         # setup optimizers after fully sharded has wrapped the lightning module
-        self.lightning_module.trainer.accelerator.setup_optimizers(self.lightning_module.trainer)
+        self.setup_optimizers(self.lightning_module.trainer)
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:160]: Neat !

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 18]
```diff
@@ -120,6 +120,16 @@ def setup(self) -> None:
         self._update_dataloader_original = pl.trainer.data_loading._update_dataloader
         pl.trainer.data_loading._update_dataloader = self._convert_to_poptorch_loader
 
+        if not self.setup_optimizers_in_pre_dispatch:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ipu.py:123]: We don't need `setup_optimizers_in_pre_dispatch` now. Maybe next PR ?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 19]
```diff
@@ -42,6 +44,9 @@ def __init__(
         checkpoint_io = checkpoint_io if checkpoint_io is not None else TorchCheckpointIO()
         self._checkpoint_io = checkpoint_io
         self._precision_plugin = precision_plugin if precision_plugin is not None else PrecisionPlugin()
+        self.optimizers: List = []
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:47]: Better typing ?

File: pytorch_lightning/plugins/training_type/sharded_spawn.py [hunk 20]
```diff
@@ -119,7 +119,7 @@ def new_process(self, trainer: "pl.Trainer", mp_queue: SimpleQueue) -> None:
         # Ensure that the scaler points to the correct process group
         # which is re-initialized in a new process
         if isinstance(self.precision_plugin, ShardedNativeMixedPrecisionPlugin):
-            self.precision_plugin.scaler = ShardedGradScaler()
+            self._precision_plugin.scaler = ShardedGradScaler()
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/sharded_spawn.py:122]:
```suggestion
            self._precision_plugin._scaler = ShardedGradScaler()
```

Missing a setter: https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=43066&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=85e03a43-2268-568b-6f18-441ae2e4d13b

File: CHANGELOG.md [hunk 21]
```diff
@@ -46,7 +46,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Moved ownership of the `PrecisionPlugin` into `TrainingTypePlugin` and updated all references ([#10570](https://github.com/PyTorchLightning/pytorch-lightning/pull/10570))
 
 
--
+- Moved optimizer related logics from `Accelerator` to `TrainingTypePlugin` ([#10596](https://github.com/PyTorchLightning/pytorch-lightning/pull/10596))
```
Comment (daniellepintz) [inline CHANGELOG.md:49]:
```suggestion
- Moved optimizer related logic from `Accelerator` to `TrainingTypePlugin` ([#10596](https://github.com/PyTorchLightning/pytorch-lightning/pull/10596))
```

File: pytorch_lightning/accelerators/accelerator.py [hunk 22]
```diff
@@ -146,16 +120,17 @@ def teardown(self) -> None:
         self.training_type_plugin.teardown()
 
     def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
-        """Moves the batch to the correct device. The returned batch is of the same type as the input batch, just
-        having all tensors on the correct device.
+        """Moves the batch to the correct device.
 
+        The returned batch is of the same type as the input batch, just
+        having all tensors on the correct device.
         Args:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:127]:
```suggestion

        Args:
```

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 15 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:67]: We already have `backend` defined above the `__init__`

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 23]
```diff
@@ -48,6 +50,14 @@ def connect(
         """Connects this plugin to the accelerator and the training process."""
         return model, optimizers, lr_schedulers
 
+    @property
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/precision_plugin.py:53]: Why do we define these properties for this class?

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 16 above)
Comment (four4fish) [reply pytorch_lightning/plugins/precision/precision_plugin.py:54]: The current logic in accelerator is 
https://github.com/PyTorchLightning/pytorch-lightning/blob/700521c7d353f79f38d1146cffc63bbe23df8e66/pytorch_lightning/accelerators/accelerator.py#L267-L273
and what pasted above. 
I think it mean to be None base on the current logic. But in refactor step 4 - "refactor accelerator_connector", we will revisit this part of the logic.

File: pytorch_lightning/plugins/training_type/ipu.py (hunk 18 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ipu.py:123]: Filed issue https://github.com/PyTorchLightning/pytorch-lightning/issues/10685

File: pytorch_lightning/trainer/trainer.py (hunk 14 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1400]: Will do!

File: pytorch_lightning/accelerators/accelerator.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:136]: @four4fish It would be fine to call l`ightning_model.trainer.accelerator.root_device` temporarily for this PR.

Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:136]: Sure, will update then!

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 24]
```diff
@@ -121,8 +121,20 @@ def pre_dispatch(self):
         if self.debug:
             os.environ["PT_XLA_DEBUG"] = str(1)
 
-    def setup(self) -> None:
+    def setup(self, trainer: "pl.Trainer") -> None:
+        # Revisit strategy inheritance
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:125]: since we have an issue for this now (#10685), we can remove the comment (assuming you meant this)

File: pytorch_lightning/lite/wrappers.py [hunk 25]
```diff
@@ -50,18 +50,20 @@ def __init__(self, optimizer: Optimizer, accelerator: Accelerator) -> None:
         self.__class__ = type("Lite" + optimizer.__class__.__name__, (self.__class__, optimizer.__class__), {})
         self._optimizer = optimizer
         self._accelerator = accelerator
+        # TODO (@awaelchli) refactor to take Strategy as param
```
Comment (awaelchli) [inline pytorch_lightning/lite/wrappers.py:53]: made a PR already: #10708

**Review by awaelchli — APPROVED** [review]
apart from existing open comments, LGTM, nice work!

File: pytorch_lightning/plugins/training_type/single_tpu.py [hunk 26]
```diff
@@ -50,14 +54,27 @@ def __init__(
...
         if is_overridden("on_post_move_to_device", self.lightning_module):
             self.model.on_post_move_to_device()
         else:
             set_shared_parameters(self.model, shared_params)
 
+        if not self.setup_optimizers_in_pre_dispatch:
+            self.setup_optimizers(trainer)
+        self.setup_precision_plugin()
+
+    def _move_optimizer_state(self, device: Optional[torch.device] = None) -> None:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/single_tpu.py:70]: the device isn't used anymore. This function should be revisited across strategies.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 24 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:125]: If kept, it's missing a `TODO:`. Also don't forget about the occurrence in `pytorch_lightning/plugins/training_type/single_tpu.py`

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 23 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/precision_plugin.py:53]: Shouldn't we remove that from the accelerator instead of adding the property here?

It's really weird to have this detail from one of the plugins leak into the common interface

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 15 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:67]: But still, what's the motivation for `amp_backend` over `backend`? Shouldn't we have just one, whichever it is?

File: pytorch_lightning/accelerators/accelerator.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:246]: added TODO in #10416 
I can make a PR for it

File: pytorch_lightning/accelerators/accelerator.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py]: updated

File: pytorch_lightning/trainer/trainer.py (hunk 14 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: updated

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 19 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py]: added!

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py:67]: backend is the outlier here, not amp_backend (the same name of the Trainer argument). We should revisit how these enums are used internally. This is touches several files outside the scope of this PR, so I suggest to do keep it like this for now. 
I will make a PR

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 23 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/precision_plugin.py]: I moved this implementation back to the existing trainer properties which is currently the main way these are accessed  everywhere. Then they don't have to be defined on the base  precision class.

File: pytorch_lightning/plugins/training_type/single_tpu.py (hunk 26 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/single_tpu.py:69]: Issue created
https://github.com/PyTorchLightning/pytorch-lightning/issues/10820

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py]: Part of the backend story is imo this simplification we need to do: https://github.com/PyTorchLightning/pytorch-lightning/issues/10821
There is my proposal how to differentiate the public attribute the user sees from what we use internally.

Comment (four4fish) [reply pytorch_lightning/plugins/precision/native_amp.py]: Oh nice! There is already a issue we can discuss the trade-offs, thanks Adrian!
